### PR TITLE
Update license to Apache License Version 2.0

### PR DESCRIPTION
**Pull request by qingyun-wu**
## Summary
This pull request changes the project license from MIT to Apache 2.0 for all new and modified code, while preserving the MIT license for original code.

### Changes Made

- Added new LICENSE file with Apache 2.0 license text
- Renamed original MIT license to LICENSE_original_MIT and included in the folder  license_original
- Added NOTICE file explaining the dual licensing
- Updated README.md to reflect license changes
- Updated file headers in modified files to Apache 2.0
- Updated the contributing guide to give contributors instructions 

### Rationale

The Apache 2.0 license was chosen for the following reasons:

- It provides explicit patent rights, which offers better protection for contributors and users
- It's compatible with many other open-source licenses, including MIT
- It's widely recognized and trusted in the open-source community

### Compliance

- Original MIT-licensed code remains under MIT license (the original license is included in the license_original/LICENSE_original_MIT file)
- All new and modified code will be under Apache 2.0
- Dual licensing is clearly explained in README and NOTICE files

### Next Steps

- Review and merge this PR
- Update project metadata on GitHub to reflect the license change
- Notify existing contributors and users about the license change

## Checks

- [x] I've included any doc changes needed for https://autogen-ai.github.io/autogen/. See https://autogen-ai.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,10 @@
 #-------------------------------------------------------------------------------------------------------------
+# Copyright (c) 2023 - 2024, Owners of https://github.com/autogen-ai
+# SPDX-License-Identifier: Apache-2.0
+# Contributions to this project, i.e., https://github.com/autogen-ai/autogen, are licensed under the Apache License, Version 2.0 (Apache-2.0).
+
+# Portions derived from  https://github.com/microsoft/autogen under the MIT License..
+# SPDX-License-Identifier: MIT
 # Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (Hk669) [inline .devcontainer/Dockerfile:8]: @qingyun-wu i think we can remove this copyrights?

File: notebook/oai_chatgpt_gpt4.ipynb [hunk 2]
```diff
@@ -17,6 +17,11 @@
...
    },
    "source": [
+    "Contributions to this project, i.e., https://github.com/autogen-ai/autogen, are licensed under the Apache License, Version 2.0 (Apache-2.0).\n",
+    "Copyright (c) 2023 - 2024, Owners of https://github.com/autogen-ai\n",
+    "SPDX-License-Identifier: Apache-2.0\n",
+    "Portions derived from  https://github.com/microsoft/autogen under the MIT License..\n",
+    "SPDX-License-Identifier: MIT\n",
     "Copyright (c) Microsoft Corporation. All rights reserved. \n",
     "\n",
     "Licensed under the MIT License.\n",
```
Comment (Hk669) [inline notebook/oai_chatgpt_gpt4.ipynb:27]: also here

File: README.md [hunk 3]
```diff
@@ -308,27 +308,13 @@ In addition, you can find:
...
 
-<!--
-# Legal Notices
-
-Microsoft and any contributors grant you a license to the Microsoft documentation and other content
-in this repository under the [Creative Commons Attribution 4.0 International Public License](https://creativecommons.org/licenses/by/4.0/legalcode),
-see the [LICENSE](LICENSE) file, and grant you a license to any code in the repository under the [MIT License](https://opensource.org/licenses/MIT), see the
-[LICENSE-CODE](LICENSE-CODE) file.
+## License
+This project is licensed under the [Apache License, Version 2.0 (Apache-2.0)](./LICENSE). We have chosen this license to ensure that future contributions are protected under terms that explicitly address patent rights and provide explicit trademark protection.
```
Comment (sonichi) [inline README.md]:
```suggestion
This project is licensed under the [Apache License, Version 2.0 (Apache-2.0)](./LICENSE).
```
You can mention the reason in a blog post or other places. Not here.

File: NOTICE.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+The [original project](https://github.com/microsoft/autogen) is licensed under the MIT License as detailed in [LICENSE_original_MIT](./license_original/LICENSE_original_MIT). The fork was created from version v0.2.35 of the original project.
+
+
+This project, i.e., https://github.com/autogen-ai/autogen, is licensed under the Apache License, Version 2.0 as detailed in [LICENSE](./LICENSE)
+
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+Ongoing MIT-licensed contributions:
+This project regularly incorporates code merged from the [original repository](https://github.com/microsoft/autogen) after the initial fork. This merged code remains under the original MIT license. For specific details on merged commits, please refer to the project's commit history.
```
Comment (sonichi) [inline NOTICE.md:16]: This looks complicated. Can we avoid the burden of checking commit history on a user?

File: notebook/oai_chatgpt_gpt4.ipynb (hunk 2 above)
Comment (qingyun-wu) [reply notebook/oai_chatgpt_gpt4.ipynb:27]: Is this comment addressed? The general principle is do not remove existing licenses.

File: NOTICE.md (hunk 4 above)
Comment (qingyun-wu) [reply NOTICE.md:16]: This is a notice. The users don't necessarily need to check the commit history. The ones who want to know details need to check.

File: notebook/oai_chatgpt_gpt4.ipynb (hunk 2 above)
Comment (Hk669) [reply notebook/oai_chatgpt_gpt4.ipynb:27]: yes let me solve this. thanks

**Review by Hk669 — APPROVED** [review]
looks good to me. thanks for the efforts in updating the license, really appreciated @qingyun-wu

**Pull request merged**